### PR TITLE
Add date_diff Presto function

### PR DESCRIPTION
**Pull request by gggrace14**
Add date_diff UDF that works the same as the corresponding Presto function, https://prestodb.io/docs/current/functions/datetime.html#interval-functions.

Supported types for date are TIMESTAMP and DATE.

### CODE REVIEW

File: velox/functions/prestosql/DateTimeImpl.h [hunk 1]
```diff
@@ -187,4 +184,120 @@ FOLLY_ALWAYS_INLINE Timestamp addToTimestamp(
...
+    return 0;
+  }
+
+  if (toDate < fromDate) {
+    return -diffDate(unit, toDate, fromDate);
+  }
+
+  return diffTimestamp(
+      unit,
+      Timestamp(fromDate.days() * 86'400, 0),
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DateTimeImpl.h:300]: Abstract this as a constant somewhere. Even using Date.h might be okay as we use this number in many places.

Comment (gggrace14) [reply velox/functions/prestosql/DateTimeImpl.h:300]: Moving those time-related constants that include kSecsPerDay from TimestampConversion.cpp to TimestampConversion.h, and include TimestampConversion.h here.

**Review by aditi-pandit — COMMENTED** [review]
Thanks for implementing this. Few comments on a quick read.

File: velox/functions/prestosql/DateTimeImpl.h [hunk 2]
```diff
@@ -187,4 +184,120 @@ FOLLY_ALWAYS_INLINE Timestamp addToTimestamp(
...
+  // TODO(gaoge): Handle overflow and underflow with 64-bit representation
+  if (fromTimestamp == toTimestamp) {
+    return 0;
+  }
+
+  if (toTimestamp < fromTimestamp) {
+    return -diffTimestamp(unit, toTimestamp, fromTimestamp);
+  }
+
+  const std::chrono::
```
[the remark this answers is not in the document]
Comment (gggrace14) [reply velox/functions/prestosql/DateTimeImpl.h:201]: Removing the recursion here diffTimestamp() & diffDate(), as inline recursive func reports error. I will change to swap the from & to in the func body.

File: velox/functions/prestosql/DateTimeImpl.h [hunk 3]
```diff
@@ -187,4 +184,120 @@ FOLLY_ALWAYS_INLINE Timestamp addToTimestamp(
...
+
+FOLLY_ALWAYS_INLINE
+int64_t
+diffDate(const DateTimeUnit unit, const Date& fromDate, const Date& toDate) {
+  if (fromDate == toDate) {
+    return 0;
+  }
+
+  if (toDate < fromDate) {
+    return -diffDate(unit, toDate, fromDate);
```
[the remark this answers is not in the document]
Comment (gggrace14) [reply velox/functions/prestosql/DateTimeImpl.h]: Removing the recursion here diffTimestamp() & diffDate(), as inline recursive func reports error.

File: velox/functions/prestosql/DateTimeImpl.h [hunk 4]
```diff
@@ -187,4 +184,120 @@ FOLLY_ALWAYS_INLINE Timestamp addToTimestamp(
...
+        return std::chrono::duration_cast<date::days>(diffMilli).count();
+      }
+      default:
+        VELOX_UNREACHABLE();
+    }
+  }
+
+  // Month, quarter and year are not uniformly incremented.
+  const date::year_month_day fromCalDate(
+      std::chrono::floor<date::days>(fromTimepoint));
```
[the remark this answers is not in the document]
Comment (gggrace14) [reply velox/functions/prestosql/DateTimeImpl.h]: Makes sense to me

**Pull request closed without merging**